replica/table.cc

```diff
@@ -521,14 +521,18 @@ compaction_group* table::single_compaction_group_if_available() const noexcept {
 }
 
 size_t table::compaction_group_idx_for_token(dht::token token) const noexcept {
-    unsigned idx = 0;
-    // avoids shift by 64.
-    if (_x_log2_compaction_groups) {
-        auto value = dht::token::to_int64(token);
-        auto mask = (1 << _x_log2_compaction_groups) - 1;
-        idx = (value >> (64 - _x_log2_compaction_groups)) & mask;
-    }
-    return idx;
+    size_t number_of_cg = 1 << _x_log2_compaction_groups;
+    size_t max_cg = number_of_cg - 1;
+    if (!_x_log2_compaction_groups || token.is_minimum()) {
+        return 0;
+    }
+    if (token.is_maximum()) {
+        return max_cg;
+    }
+    int64_t value = dht::token::to_int64(token);
+    uint8_t log2_shift = (64 - _x_log2_compaction_groups);
+    int64_t mask = int64_t(max_cg) << log2_shift;
+    return ((value & mask) >> log2_shift) + number_of_cg / 2;
```

Comment:
I think it will be simpler if you bias the token before shifting. See dht::shard_of().

Reply:
I think we can even reuse `shard_of(unsigned shard_count, unsigned sharding_ignore_msb_bits, const token& t)` as is.
Need to measure the overhead of uint128_t arithmetic there.  I'm not sure how its performance compares to the code above (on x86_64 or other architectures).

Reply:
No, sorry, we can't reuse it as is as in this case the "sharding" into compaction group is based on the most-significant bits rather than ignoring them, but the logic is indeed similar.

Reply:
changed in v2 / #12599